fendermint/app/src/cmd/upgrades/upgrade02.rs

```diff
@@ -1,50 +1,85 @@
+                "0x00000000000000000000000000000000000000000000000000000000000000030000000000000000000000007058755aee20c1f170de6300bb9bb177861a9a440000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000006a00000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000186a000000000000000000000000000000000000000000000000000000000000000410435e9b95277b1b41516aec360163cc8e041e715a5d81c1a0e6a17d40365ea0a25d87a058d6164c348b2fa3cd55710d3b41943f408c6301918717a68557231c8bf00000000000000000000000000000000000000000000000000000000000000",
+            ];
+
+            let logs = parse_logs(raw_configuration_changes)?;
+
+            let validator_changes = ethers_contract::decode_logs::<NewStakingChangeRequestFilter>(&logs)?
+                .into_iter()
+                .map(|p| {
+                    StakingChangeRequest{
+                        configuration_number: p.configuration_number,
+                        change: StakingChange {
+                            op: p.op,
+                            payload: p.payload,
+                            validator: p.validator
+                        }
+                    }
+                })
+                .collect();
+
+            let gateway_addr = EthAddress::from(
+                ethers::types::Address::from_str("0x77aa40b105843728088c0132e43fc44348881da8")
+                    .expect("invalid gateway addr"),
+            );
 
             info!(
-                "[Upgrade at height {}] Change gateway ownership",
+                "[Upgrade at height {}] Apply missing validator changes",
                 state.block_height()
             );
 
-            let ownership = ContractCaller::<_, _, OwnershipFacetErrors>::new(
-                EthAddress::from(gateway_addr),
-                OwnershipFacet::new,
+            let topdown = ContractCaller::<_, _, TopDownFinalityFacetErrors>::new(
+                gateway_addr,
+                TopDownFinalityFacet::new,
+            );
+
+            let ret = topdown.call_with_return(state, |c| c.store_validator_changes(validator_changes))?.into_return();
+
+            info!(
+                "[Upgrade at height {}] Applied validator changes with result: {:#?}",
+                state.block_height(),
+                ret.apply_ret.msg_receipt,
```

Comment:
Let's also print number of applied events, so we can always check it's correct.

```suggestion
            info!(
                "[Upgrade at height {}] Applied {} validator changes with result: {:#?}",
                state.block_height(),
                validator_changes.len(),
                ret.apply_ret.msg_receipt,
```

Would also be great to retrieve final Validator Configuration Number after these events. Is that possible? :)